### PR TITLE
Allow to byteswap data when reading saved torch jit data

### LINKED ISSUE #151428: TorchScript Model Saved on x86 Returns NaNs When Loaded on s390x
### 🐛 Describe the bug

I trained a simple LSTM model on the Iris dataset using PyTorch on an **x86 (little-endian)** system. I then saved the model using both `torch.jit.trace()` and `torch.jit.script()`.

When loading this `.pt` scripted model on an **s390x (big-endian)** architecture, the model loads **without error**, but the output is **entirely NaN**, even when using `set_default_load_endianness(LoadEndianness.LITTLE)` before loading.

This issue does **not occur** when loading the model on x86 again. Only the s390x platform exhibits this behavior.

**Reproduction Script**

- Save the model on x86_64
```
import torch
import torch.nn as nn
import torch.nn.functional as F
from sklearn.datasets import load_iris
from sklearn.preprocessing import StandardScaler
# Set deterministic behavior
torch.manual_seed(0)
# Define a simple LSTM model
class LSTMClassifier(nn.Module):
    def __init__(self, input_size, hidden_size, num_classes):
        super(LSTMClassifier, self).__init__()
        self.lstm = nn.LSTM(input_size, hidden_size, batch_first=True)
        self.fc = nn.Linear(hidden_size, num_classes)
def forward(self, x):
        x = x.unsqueeze(1)  # (batch, seq_len=1, input_size)
        lstm_out, _ = self.lstm(x)
        return self.fc(lstm_out[:, -1, :])  # logits
# Load data
X = load_iris().data
X = StandardScaler().fit_transform(X)
X_tensor = torch.tensor(X, dtype=torch.float32)
# Train dummy model
model = LSTMClassifier(input_size=4, hidden_size=16, num_classes=3)
criterion = nn.CrossEntropyLoss()
optimizer = torch.optim.Adam(model.parameters(), lr=0.01)
y = torch.randint(0, 3, (150,))
for epoch in range(10):
    out = model(X_tensor)
    loss = criterion(out, y)
    optimizer.zero_grad()
    loss.backward()
    optimizer.step()
# Save with torch.jit.script
scripted_model = torch.jit.script(model)
torch.jit.save(scripted_model, "lstm_script_model.pt")
# Save with torch.jit.trace
traced_model = torch.jit.trace(model, torch.randn(1, 4))
torch.jit.save(traced_model, "lstm_traced_model.pt")
```

- Load the model on s390x (zLinux)
```
import torch
from torch.serialization import set_default_load_endianness
from torch.utils.serialization.config import LoadEndianness
# Set little-endian since model was saved on x86
set_default_load_endianness(LoadEndianness.LITTLE)
# Load the TorchScript model
model = torch.jit.load("lstm_script_model.pt")
model.eval()
# Test input
input_data = torch.tensor([
    [0.1, -0.2, 0.3, 0.4],
    [-1.1, 0.9, -0.3, 0.7],
    [1.2, -1.2, 1.3, -1.1],
    [0.3, 0.5, -0.4, 0.2]
], dtype=torch.float32)
# Inference
with torch.no_grad():
    output = model(input_data)
print("Output:")
print(output)

# Load the TorchScript trace model
model = torch.jit.load("lstm_trace_model.pt")
model.eval()
# Inference
with torch.no_grad():
    output = model(input_data)
print("Output:")
print(output)
```

**Observed Behavior**
```
tensor([[nan, nan, nan],
        [nan, nan, nan],
        [nan, nan, nan],
        [nan, nan, nan]])
```
Even though the model loads successfully and is in eval() mode, the inference result is NaN across all outputs.

**Expected Behavior**
The model should return consistent logits as it does on x86.
```
tensor([[-4.2518,  6.0927, -1.8566],
        [ 6.4866, -2.5016, -3.5421],
        [-3.9846, -3.5373,  6.0734],
        [-4.0324,  4.9734, -1.0101]])
```

**Request**
Would appreciate help in:
- Is this a known TorchScript limitation across different architectures (x86 vs s390x)?
- Are there any workarounds (e.g., manually loading LSTM weights)?
- Could this be a bug in how TorchScript handles LSTM/any other model's weight serialization across architectures?

### Versions

```
PyTorch version: 2.5.0
Is debug build: False
CUDA used to build PyTorch: None
ROCM used to build PyTorch: N/A

OS: Ubuntu 22.04.5 LTS (s390x)
GCC version: (Ubuntu 11.4.0-1ubuntu1~22.04) 11.4.0
Clang version: Could not collect
CMake version: Could not collect
Libc version: glibc-2.35

Python version: 3.10.12 (main, Feb  4 2025, 14:57:36) [GCC 11.4.0] (64-bit runtime)
Is CUDA available: False
CUDA runtime version: No CUDA
CUDA_MODULE_LOADING set to: N/A
GPU models and configuration: No CUDA
Nvidia driver version: No CUDA
cuDNN version: No CUDA
HIP runtime version: N/A
MIOpen runtime version: N/A
Is XNNPACK available: False

CPU:
Architecture:                         s390x
CPU op-mode(s):                       32-bit, 64-bit
Byte Order:                           Big Endian

Versions of relevant libraries:
[pip3] numpy==1.26.4
[pip3] torch==2.5.0
```

### PR DESCRIPTION
**Pull request by AlekseiNikiforovIBM**
It looks like some pickled data is endian-dependent. Byteswap such data when needed.

Add testcases.

Fixes #151428 

cc @EikanWang @jgong5 @wenzhe-nrv @sanchitintel

### CODE REVIEW

**Review by malfet — CHANGES_REQUESTED** [review]
Please do not embed large files into .py, just store them as artifacts (it looks like you are adding nearly 1500 lines to test_save_load.py)

**Comment by AlekseiNikiforovIBM** [comment]
Sure, I'll update tests.

**Comment by AlekseiNikiforovIBM** [comment]
Tests are updated and build is fixed. Please take a look.

**Comment by AlekseiNikiforovIBM** [comment]
Could you please take a look at this PR again?

**Comment by AlekseiNikiforovIBM** [comment]
This change is still relevant.

**Comment by AlekseiNikiforovIBM** [comment]
This PR is still relevant. Could you please review it?

**Comment by albanD** [comment]
Thank you very much for sending a PR.
To avoid you spending time preparing a PR for a feature we're not sure we want to add, our usual process is to discuss on the issue before doing the PR.
See https://github.com/pytorch/pytorch/wiki/The-Ultimate-Guide-to-PyTorch-Contributions for details!

In this particular case, I'm afraid no one is working on TorchScript so I'm not sure if anyone will fix this. Maybe @SherlockNoMad if this API is still used on your end?

**Comment by AlekseiNikiforovIBM** [comment]
The PR addresses an issue reported by a user. It is not adding any new functionality. Rather, it closes a gap for an otherwise well-supported platform. The code does not pose any risk to other platforms. The code is ready to be merged and includes test cases to ensure it continues to function properly.

**Comment by AlekseiNikiforovIBM** [comment]
I've ran following small piece of code on x86_64:
```
#!/usr/bin/env python3

import torch

for i in range(0,100000):
        models = torch.jit.load("file.pt")
        print(i)
        print(models.fc.bias)
```

where `file.pt` was `test/jit/serialized/TestSaveLoadFlatbuffer.test_save_load_byteorder_be.input.pt` or `test/jit/serialized/TestSaveLoadFlatbuffer.test_save_load_byteorder_le.input.pt` from added tests.

I've ran it with this change and without it.
In all cases performance was similar, without any noticeable differences.
